src/python/gudhi/sklearn/rips_persistence.py

```diff
@@ -73,36 +84,90 @@ def fit(self, X, Y=None):
+
+                input_type = 'distance coo_matrix'
+            else:
+                inp = squareform(pdist(inp))
+                input_type = 'full distance matrix'
+
+        # Dense -> sparse
+        if input_type in ('full distance matrix', 'lower distance matrix'):
+            # After this filtration value, all complexes are cones, nothing happens
+            if input_type == 'full distance matrix':
+                inp = np.asarray(inp)
+                cone_radius = inp.max(-1).min()
+            else:
+                cone_radius = _lower_cone_radius(inp)
+            sparsify = use_simplex_tree or num_collapses > 0 or threshold < cone_radius # heuristic
+            threshold = min(threshold, cone_radius)
+            if sparsify:
+                # For 'full' we could use i, j = np.triu_indices_from(inp, k=1), etc
+                i, j, f = _lower_to_coo(inp, threshold)
+                inp = coo_matrix((f, (i, j)), shape=(n,) * 2)
+                input_type = 'distance coo_matrix'
+
+        if num_collapses > 0:
+            assert input_type == 'distance coo_matrix'
+            inp = reduce_graph(inp, num_collapses)
+
+        if use_simplex_tree:
+            from gudhi import SimplexTree
+            st = SimplexTree()
+            # Use create_from_array in case of full matrix?
+            # (not important since this fallback mostly matters in high dimension, where we use edge-collapse anyway)
+            st.insert_batch(np.arange(n).reshape(1,-1), np.zeros(n))
+            st.insert_edges_from_coo_matrix(inp)
+            st.expansion(max_dimension + 1)
+            st.compute_persistence(homology_coeff_field=self.homology_coeff_field, persistence_dim_max=max_dimension>=st.dimension())
+            return [ st.persistence_intervals_in_dimension(dim) for dim in self.dim_list_ ]
+
+        if input_type == 'full distance matrix':
+            ## Possibly transpose for performance?
+            # if inp.strides[0] > inp.strides[1]: # or the reverse?
+            #     inp = inp.T
+            dgm = _full(inp, max_dimension=max_dimension, max_edge_length=threshold, homology_coeff_field=self.homology_coeff_field)
+        elif input_type == 'lower distance matrix':
+            dgm = _lower(inp, max_dimension=max_dimension, max_edge_length=threshold, homology_coeff_field=self.homology_coeff_field)
+        elif input_type == 'distance coo_matrix':
+            # Switch to coo_array (danger: row/col seem deprecated)?
+            dgm = _sparse(inp.row, inp.col, inp.data, inp.shape[0], max_dimension=max_dimension, max_edge_length=threshold, homology_coeff_field=self.homology_coeff_field)
         else:
-            raise ValueError("Only 'point cloud' and  'lower distance matrix' are valid input_type")
+            raise ValueError("Only 'point cloud', 'lower distance matrix', 'full distance matrix' and 'distance coo_matrix' are valid input_type") # move to __init__?
```

Comment:
From a user point of view, I think it is better to catch exception at construction

Reply:
`input_type` could have been changed by `set_params`, so I don't think the constructor is sufficient, but I guess we could have the test in 2 places.

Reply:
https://scikit-learn.org/stable/developers/develop.html#instantiation says
> To summarize, an `__init__` should look like: [...] There should be no logic, not even input validation,

I am ok with violating this recommendation if you think we should, but I wanted to get confirmation first.

Reply:
Ah... I was not aware of these recommendations, sorry about that. We should decide if we follow it or not and explain it in the [code_conventions](https://github.com/GUDHI/gudhi-devel/blob/master/.github/code_conventions.md). Not for this PR.

Reply:
We did discuss them a few times already :wink:
It led to some changes in the representations module.
I think we should *try* to follow it (if we don't, something may mysteriously fail deep in a pipeline), but not dogmatically:
* If we decide to do some argument validation in `__init__`, I don't really see how that can break anything.
* Currently we don't have a default value for homology_dimensions, and I am ok with that. If it causes problems, the error message is likely to be easy to understand, and we can consider adding a default later.